File: skills/load-API-key/references/shared_apikey_loader.py
"""
shared_apikey_loader.py
========================
부서 공용서버에 위치한 .env 파일에서 API 키를 로드하는 유틸리티.

사용법:
    from shared_apikey_loader import load_shared_keys, get_key

    load_shared_keys()
    api_key = get_key("OPENAI_API_KEY")

설정:
    settings.ini 파일에서 공용서버 경로를 관리합니다.
    settings.ini는 이 스크립트와 같은 디렉토리에 위치해야 합니다.

의존성:
    pip install python-dotenv
"""

import os
import configparser
from pathlib import Path
from dotenv import load_dotenv


# ---------------------------------------------------------------------------
# 설정 로드
# ---------------------------------------------------------------------------
def _get_settings_path() -> Path:
    """settings.ini 경로를 반환한다. 스크립트와 같은 디렉토리에서 찾는다."""
    return Path(__file__).parent / "settings.ini"


def _read_settings() -> configparser.ConfigParser:
    """settings.ini를 읽어 ConfigParser 객체로 반환한다."""
    settings_path = _get_settings_path()
    if not settings_path.exists():
        raise FileNotFoundError(
            f"설정 파일을 찾을 수 없습니다: {settings_path}\n"
            f"settings.ini 파일을 생성한 뒤 공용서버 경로를 입력하세요."
        )
    config = configparser.ConfigParser()
    config.read(str(settings_path), encoding="utf-8")
    return config


# ---------------------------------------------------------------------------
# 키 로드
# ---------------------------------------------------------------------------
def load_shared_keys() -> bool:
    """
    settings.ini에 지정된 공용서버 .env 파일을 읽어 환경변수로 로드한다.

    Returns:
        True  - 공용서버 .env 로드 성공
        False - 폴백(.env 로컬) 사용 또는 실패
    """
    config = _read_settings()

    env_path = config.get("server", "env_path", fallback="")
    encoding = config.get("options", "encoding", fallback="utf-8")
    use_fallback = config.getboolean("options", "use_local_fallback", fallback=True)
    local_env = config.get("options", "local_env_path", fallback=".env")

    # 1) 공용서버 경로 시도
    if env_path and Path(env_path).is_file():
        load_dotenv(dotenv_path=env_path, encoding=encoding, override=True)
        print(f"[INFO] 공용서버 .env 로드 완료: {env_path}")
        return True

    print(f"[WARN] 공용서버 .env 접근 불가: {env_path}")

    # 2) 로컬 폴백
    if use_fallback:
        local_path = Path(local_env) if Path(local_env).is_absolute() else Path.cwd() / local_env
        if local_path.is_file():
            load_dotenv(dotenv_path=str(local_path), encoding=encoding, override=True)
            print(f"[INFO] 로컬 폴백 .env 로드 완료: {local_path}")
            return False

    print("[ERROR] 사용 가능한 .env 파일이 없습니다.")
    return False


def get_key(key_name: str, default: str | None = None) -> str | None:
    """
    환경변수에서 API 키를 가져온다.

    Args:
        key_name: 환경변수 이름 (예: 'OPENAI_API_KEY')
        default:  키가 없을 때 반환할 기본값

    Returns:
        API 키 문자열 또는 default
    """
    value = os.getenv(key_name, default)
    if value is None:
        print(f"[WARN] 키를 찾을 수 없습니다: {key_name}")
    return value


def list_keys() -> dict[str, str]:
    """
    현재 로드된 API 키 목록을 반환한다.
    키 값은 앞 8자만 표시하고 나머지는 마스킹한다.
    """
    config = _read_settings()
    env_path = config.get("server", "env_path", fallback="")

    keys = {}
    # .env 파일을 직접 파싱하여 키 이름 목록 확보
    for path in [env_path, config.get("options", "local_env_path", fallback=".env")]:
        p = Path(path)
        if p.is_file():
            with open(p, encoding="utf-8") as f:
                for line in f:
                    line = line.strip()
                    if line and not line.startswith("#") and "=" in line:
                        name = line.split("=", 1)[0].strip()
                        val = os.getenv(name, "")
                        if val:
                            masked = val[:8] + "*" * max(0, len(val) - 8)
                            keys[name] = masked
            break
    return keys


# ---------------------------------------------------------------------------
# 직접 실행 시 테스트
# ---------------------------------------------------------------------------
if __name__ == "__main__":
    print("=" * 50)
    print("공용서버 API 키 로더 테스트")
    print("=" * 50)

    load_shared_keys()

    print("\n--- 로드된 키 목록 (마스킹) ---")
    for name, masked in list_keys().items():
        print(f"  {name}: {masked}")

    print("\n--- 개별 키 조회 예시 ---")
    openai_key = get_key("OPENAI_API_KEY")
    anthropic_key = get_key("ANTHROPIC_API_KEY")
    print(f"  OPENAI_API_KEY    = {openai_key}")
    print(f"  ANTHROPIC_API_KEY = {anthropic_key}")

File: skills/load-API-key/references/SharedApiKeyLoader.cs
/*
 * SharedApiKeyLoader.cs
 * =====================
 * 부서 공용서버에 위치한 .env 파일에서 API 키를 로드하는 유틸리티.
 *
 * 사용법:
 *     var loader = new SharedApiKeyLoader();       // settings.ini 자동 탐색
 *     loader.LoadSharedKeys();
 *     string apiKey = loader.GetKey("OPENAI_API_KEY");
 *
 * 의존성: 없음 (표준 라이브러리만 사용)
 */

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

/// <summary>
/// settings.ini에 지정된 공용서버 .env 파일에서 API 키를 로드한다.
/// </summary>
public class SharedApiKeyLoader
{
    private readonly string _settingsPath;
    private readonly Dictionary<string, string> _keys = new();

    /// <param name="settingsPath">
    /// settings.ini 경로. null이면 실행 파일과 같은 디렉토리에서 자동 탐색.
    /// </param>
    public SharedApiKeyLoader(string settingsPath = null)
    {
        _settingsPath = settingsPath
            ?? Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "settings.ini");
    }

    /// <summary>
    /// settings.ini에 지정된 공용서버 .env 파일을 읽어 내부 딕셔너리에 로드한다.
    /// </summary>
    /// <returns>공용서버 .env 로드 성공 여부</returns>
    public bool LoadSharedKeys()
    {
        if (!File.Exists(_settingsPath))
            throw new FileNotFoundException($"설정 파일을 찾을 수 없습니다: {_settingsPath}");

        var ini = ReadIni(_settingsPath);

        string envPath = GetIniValue(ini, "server", "env_path", "");
        string encoding = GetIniValue(ini, "options", "encoding", "utf-8");
        bool useFallback = GetIniValue(ini, "options", "use_local_fallback", "true")
            .Equals("true", StringComparison.OrdinalIgnoreCase);
        string localEnvPath = GetIniValue(ini, "options", "local_env_path", ".env");

        var enc = Encoding.GetEncoding(encoding);

        // 1) 공용서버 경로 시도
        if (!string.IsNullOrEmpty(envPath) && File.Exists(envPath))
        {
            LoadEnvFile(envPath, enc);
            Console.WriteLine($"[INFO] 공용서버 .env 로드 완료: {envPath}");
            return true;
        }

        Console.WriteLine($"[WARN] 공용서버 .env 접근 불가: {envPath}");

        // 2) 로컬 폴백
        if (useFallback)
        {
            string localPath = Path.IsPathRooted(localEnvPath)
                ? localEnvPath
                : Path.Combine(Directory.GetCurrentDirectory(), localEnvPath);

            if (File.Exists(localPath))
            {
                LoadEnvFile(localPath, enc);
                Console.WriteLine($"[INFO] 로컬 폴백 .env 로드 완료: {localPath}");
                return false;
            }
        }

        Console.WriteLine("[ERROR] 사용 가능한 .env 파일이 없습니다.");
        return false;
    }

    /// <summary>API 키를 가져온다.</summary>
    public string GetKey(string keyName, string defaultValue = null)
    {
        if (_keys.TryGetValue(keyName, out string value))
            return value;

        // 환경변수에서도 시도
        string envValue = Environment.GetEnvironmentVariable(keyName);
        if (envValue != null)
            return envValue;

        if (defaultValue == null)
            Console.WriteLine($"[WARN] 키를 찾을 수 없습니다: {keyName}");

        return defaultValue;
    }

    /// <summary>로드된 키 목록을 마스킹하여 반환한다.</summary>
    public Dictionary<string, string> ListKeys()
    {
        return _keys.ToDictionary(
            kv => kv.Key,
            kv => kv.Value.Length > 8
                ? kv.Value[..8] + new string('*', kv.Value.Length - 8)
                : kv.Value
        );
    }

    // ---------------------------------------------------------------
    // 내부 메서드
    // ---------------------------------------------------------------

    private void LoadEnvFile(string path, Encoding encoding)
    {
        foreach (string rawLine in File.ReadAllLines(path, encoding))
        {
            string line = rawLine.Trim();
            if (string.IsNullOrEmpty(line) || line.StartsWith("#"))
                continue;

            int idx = line.IndexOf('=');
            if (idx <= 0) continue;

            string key = line[..idx].Trim();
            string val = line[(idx + 1)..].Trim().Trim('"');
            _keys[key] = val;
            Environment.SetEnvironmentVariable(key, val);
        }
    }

    /// <summary>간이 INI 파서. [section] → key=value 구조를 딕셔너리로 반환.</summary>
    private static Dictionary<string, Dictionary<string, string>> ReadIni(string path)
    {
        var result = new Dictionary<string, Dictionary<string, string>>(
            StringComparer.OrdinalIgnoreCase);
        string currentSection = "";

        foreach (string rawLine in File.ReadAllLines(path, Encoding.UTF8))
        {
            string line = rawLine.Trim();
            if (string.IsNullOrEmpty(line) || line.StartsWith("#") || line.StartsWith(";"))
                continue;

            if (line.StartsWith("[") && line.EndsWith("]"))
            {
                currentSection = line[1..^1].Trim();
                if (!result.ContainsKey(currentSection))
                    result[currentSection] = new Dictionary<string, string>(
                        StringComparer.OrdinalIgnoreCase);
                continue;
            }

            int idx = line.IndexOf('=');
            if (idx > 0 && result.ContainsKey(currentSection))
            {
                string key = line[..idx].Trim();
                string val = line[(idx + 1)..].Trim();
                result[currentSection][key] = val;
            }
        }

        return result;
    }

    private static string GetIniValue(
        Dictionary<string, Dictionary<string, string>> ini,
        string section, string key, string fallback)
    {
        if (ini.TryGetValue(section, out var sec) && sec.TryGetValue(key, out string val))
            return val;
        return fallback;
    }
}

// ---------------------------------------------------------------------------
// 사용 예제 (콘솔 앱에서 직접 실행)
// ---------------------------------------------------------------------------
// class Program
// {
//     static void Main()
//     {
//         var loader = new SharedApiKeyLoader();
//         loader.LoadSharedKeys();
//
//         // 키 목록 (마스킹)
//         foreach (var kv in loader.ListKeys())
//             Console.WriteLine($"  {kv.Key}: {kv.Value}");
//
//         // 개별 키 참조
//         string openaiKey = loader.GetKey("OPENAI_API_KEY");
//         string anthropicKey = loader.GetKey("ANTHROPIC_API_KEY");
//     }
// }
